Fix issue #6872 - Mutex lock has possibility to fail at runtime (returning status flag)

mprse opened the pull request:
### Description

This patch is a partial fix for issue https://github.com/ARMmbed/mbed-os/issues/6872.

This is the first step which adds an assertion, so we get the indication in develop builds, not just debug.
Potentially in the next step we could perform total elimination of the error return in the non-time-limited Mutex claim.

More information can be found here:
https://jira.arm.com/browse/IOTMORF-2350


### Pull request type

    [ ] Fix
    [ ] Refactor
    [ ] New target
    [ ] Feature
    [X] Breaking change

mprse committed:
Fix for issue #6872 - Mutex lock has possibility to fail at runtime (returning status flag)

Add assertions for statuses returned by osMutexAcquire().
Add void Mutex::lock(void) member function - lock function which does not return status and waits for the mutex forever.
Make osStatus Mutex::lock(uint32_t millisec=osWaitForever) member function deprecated in favour of bool Mutex::trylock_for(uint32_t millisec).

kjbracey commented on rtos/Mutex.cpp:
```diff
@@ -55,6 +55,9 @@ osStatus Mutex::lock(uint32_t millisec) {
         _count++;
     }
+
+    MBED_ASSERT(status == osErrorTimeout || status == osOK);
```
Unfortunately CMSIS-RTOS can also return "osErrorResource" if the mutex is locked and timeout is 0.

And there should be no error if the timeout specified was 0xffffffff. I think.

The maximally-tight assertion is I think:

    MBED_ASSERT(status == osOK ||
                (status == osErrorResource && millisec == 0) ||
                (status == osErrorTimeout && millisec != osWaitForever));
 
Note that if adding an assert here, `trylock_for` can have its removed.

kjbracey reviewed (commented):
I had forgotten that `trylock_for` had made it in already.

Given that that is present, it might also make sense to add an overload for `lock(void)` (which is the one which ultimately will get a void return) and deprecate `lock(millisec)` in favour of `trylock_for`.

mprse committed:
Fix codding style.

mprse committed:
`osStatus lock(uint32_t millisec=osWaitForever)` - remove default value

Since `void lock(void)` member function has been added we need to remove default value from deprecated `lock(uint32_t millisec=osWaitForever)` member function since otherwise calling `lock()` would be ambiguous.

mprse committed:
Fix usage of `Mutex::lock()` function - ignore returned value.

mprse committed:
Adapt `tests-mbedmicro-rtos-mbed-mutex` test - use new lock(void) function instead of deprecated one.

mprse commented:
@kjbracey-arm Thanks for the review. I followed your suggestion and added `void Mutex::lock(void)`  member function and marked `osStatus lock(uint32_t millisec=osWaitForever);` as deprecated.
As a result of these operations the following additional changes were required:
- remove default value from deprecated `lock(uint32_t millisec=osWaitForever)` member function since otherwise calling `mutex.lock()` would be ambiguous.
- fix usage of `Mutex::lock()` function is source file - ignore returned value.
- adapt `tests-mbedmicro-rtos-mbed-mutex` test - use new `void lock(void)` function instead of deprecated one.

The following questions appeared:
- Which mbed version should be used in `MBED_DEPRECATED_SINCE` macro? I used `mbed-os-5.10.0` ?
- Now we have inconsistency between `lock()` and `unlock()`. Maybe the same modifications should be performed for `Mutex::unlock()` member function?

kjbracey commented:
I was thinking it was premature to actually make the `Mutex::lock()` return `void` - suggesting the separation only to get the deprecation warning on the `(millisecond)` form - but if that few people really are checking the return value in-tree, maybe it's actually feasible. May actually lead to fewer problems overall, especially with binary compatibility, if that overload has `void` return from the beginning. I'll approve if others are fine with it.

Deprecated-since should say 5.10.

Yes, when `Mutex::lock()` is changed to `void` return, so should `unlock` be. And it should also assert either way.

mprse committed:
Change `Mutex::unlock()` function definition - void return value

This is done for consistency with the new version of `Mutex::lock()` member function which does not return status.

mprse committed:
Fix usage of `Mutex::unlock()` function - ignore returned value.

mprse committed:
Adapt `tests-mbedmicro-rtos-mbed-mutex` test - use new void unlock(void) function

mprse commented:
Changed definition of `Mutex::unlock` member function to `void unlock(void);` for consistency with `void lock(void);`.

0xc0170 commented:
> [ ] Breaking change

Isn't this breaking change? I dont see any mention in the commit message but from the discussion and code, looks like it is. If yes, this will need careful consideration and review

mprse commented:
> Isn't this breaking change? I dont see any mention in the commit message but from the discussion and code, looks like it is. If yes, this will need careful consideration and review

Marked this PR as "breaking change".

mprse commented:
@ARMmbed/mbed-os-core ping

bulislaw commented:
ping

deepikabhavnani commented on TESTS/mbedmicro-rtos-mbed/mutex/main.cpp:
```diff
@@ -224,46 +219,40 @@ void test_dual_thread_lock(void)
 
-    stat = mutex.lock();
-    TEST_ASSERT_EQUAL(osOK, stat);
+    mutex.lock();
```
We don;t return any status, but how do we verify if recursive lock here was successful or not?

mprse replied:
In the original version returned status has been checked. Since we do not have status now we can only execute two successive locks and two successive unlocks and show that the test case has been executed successfully. We do not have status, but please note that ASSERT has been added to the lock/unlock. In case of failure ASSERT will force the test to fail.

deepikabhavnani commented on TESTS/mbedmicro-rtos-mbed/mutex/main.cpp:
```diff
@@ -272,26 +261,23 @@ void test_single_thread_trylock(void)
 
-    stat = mutex.unlock();
-    TEST_ASSERT_EQUAL(osOK, stat);
+    mutex.unlock();
```
I am not sure how are we testing lock/unlock success/failure here? Should we modify the test to check from another thread if lock was applied or not and set success/fail based on that.

mprse replied:
ASSERT added in lock/unlock will force the test to fail in case of failure. 
I don't think that we need here to use another thread to show that lock/unlock works since we got other test cases which proves that.

deepikabhavnani reviewed (approved):
👍

cmonr commented:
Once this is rebased, we can start CI.

mprse commented:
> Once this is rebased, we can start CI.

Rebased and after consultation with @0xc0170 marked `osStatus lock(uint32_t millisec)` as deprecated since `mbed-os-6.0`.

cmonr commented:
/morph build

cmonr commented:
@kjbracey-arm @deepikabhavnani @ARMmbed/mbed-os-core @bulislaw @0xc0170 
Would anyone care to take a final look?

kjbracey commented:
I think the deprecation of `lock(millisec)` is for now - 5.10. We know we want people to move to `lock()` or `try_lock(millisec)`.

The main question is what the return type of `lock()` should be, which doesn't affect the above deprecation. If it returns `void`, which is what we ultimately want, it's a breaking change for anyone checking the return value without timeout.

But we can't deprecate just use of the return value. :(  Either we do this now despite lack of deprecation or save it to some 6.0 point, where we'd be okay with a change without deprecation.

I don't want to sidestep by use of a `lock2()` or something, because we do ultimately want to get the point of having just `void lock()` to be a C++ `Lockable`.

0xc0170 commented:
> I think the deprecation of lock(millisec) is for now - 5.10. We know we want people to move to lock() or try_lock(millisec).

@mprse Please correct that

> But we can't deprecate just use of the return value. :( Either we do this now despite lack of deprecation or save it to some 6.0 point, where we'd be okay with a change without deprecation.

I would vote the next major.

kjbracey commented:
I'd be fine with making it `void` on next major if that's the consensus. An unnecessary return of `osStatus` on a "can't fail" `lock()` doesn't actually really matter, as long as it's documented that the call asserts before returning anything other than OK. We would still be a `Lockable` - the requirement is just that we have a `lock()`, the return doesn't matter. Making it `void` is just about neatness, really.

deepikabhavnani reviewed (approved):
Looks good.

mprse commented:
> I think the deprecation of lock(millisec) is for now - 5.10. We know we want people to move to lock() or try_lock(millisec).
> 
> @mprse Please correct that

Reverted to 5.10.

cmonr commented:
Export failure looked like it was all over the place.

Going to rebuild, but since the merge checks reset, it looks like something happened with the PR.

/morph build

mprse commented:
```
/builds/ws/exporter-build-matrix/2399/ARCH_PRO_make_gcc_arm/sources/examples/mbed-os-example-blinky/Makefile:1131: recipe for target 'mbed-os/features/device_key/source/DeviceKey.o' failed
Makefile:25: recipe for target 'all' failed
==========STDERR==========
In file included from .././mbed-os/platform/PlatformMutex.h:29:0,
                 from .././mbed-os/features/nvstore/source/nvstore.h:30,
                 from ../mbed-os/features/device_key/source/DeviceKey.cpp:20:
.././mbed-os/rtos/Mutex.h:103:27: error: expected identifier before string constant
     MBED_DEPRECATED_SINCE("mbed-os-5.10.0", "Replaced with trylock_for and lock(void) functions")
                           ^~~~~~~~~~~~~~~~
.././mbed-os/rtos/Mutex.h:103:27: error: expected ',' or '...' before string constant
.././mbed-os/rtos/Mutex.h:103:97: error: ISO C++ forbids declaration of 'MBED_DEPRECATED_SINCE' with no type [-fpermissive]
     MBED_DEPRECATED_SINCE("mbed-os-5.10.0", "Replaced with trylock_for and lock(void) functions")
                                                                                                 ^
.././mbed-os/rtos/Mutex.h:103:97: error: expected ';' at end of member declaration
In file included from .././mbed-os/platform/PlatformMutex.h:29:0,
                 from ../mbed-os/drivers/FlashIAP.h:29,
                 from ../mbed-os/drivers/FlashIAP.cpp:26:
.././mbed-os/rtos/Mutex.h:103:27: error: expected identifier before string constant
     MBED_DEPRECATED_SINCE("mbed-os-5.10.0", "Replaced with trylock_for and lock(void) functions")
                           ^~~~~~~~~~~~~~~~
.././mbed-os/rtos/Mutex.h:103:27: error: expected ',' or '...' before string constant
.././mbed-os/rtos/Mutex.h:103:97: error: ISO C++ forbids declaration of 'MBED_DEPRECATED_SINCE' with no type [-fpermissive]
     MBED_DEPRECATED_SINCE("mbed-os-5.10.0", "Replaced with trylock_for and lock(void) functions")
                                                                                                 ^
.././mbed-os/rtos/Mutex.h:103:97: error: expected ';' at end of member declaration
make[1]: *** [mbed-os/drivers/FlashIAP.o] Error 1
```

mprse committed:
Mutex: fixing missing include

Macro MBED_DEPRECATED_SINCE is defined in platform/mbed_toolchain.h which was not included.
If someone used member functions lock (which are prefixed with MBED_DEPRECATED_SINCE since some time), there would be a compile error instead of a warning.
Including mbed_toolchain.h fixes that.

mprse commented:
Added missing include. This should solve the problem with the exporter builds.

kjbracey commented on rtos/Mutex.cpp:
```diff
@@ -50,30 +51,51 @@ void Mutex::constructor(const char *name)
 }
 
-osStatus Mutex::lock(uint32_t millisec) {
+void Mutex::lock(void)
```
Need a final decision on this return type, as per previous comment. @deepikabhavnani - any views?

0xc0170 replied:
I thought we agreed no breaking change, thus revert this one, and add documentation note about return

kjbracey replied:
I'm fine with that, but still have a slight preference for just voiding it, so looking for 3rd or 4th votes just to make sure...

mprse replied:
Guys I'm a little confused now. So do you want me to revert everything and just add assertions to the `lock/unlock` functions and documentation note about the asserts?

kjbracey replied:
The proposal was to have this separate `lock()` overload with the assert, but have it returning `osStatus` to minimise the compilation breakage. Documentation should note that it asserts and that use of the return value is deprecated, and it's planned to become void in 6.0. But that would have to be just as a note - you couldn't do it as a "MBED_DEPRECATED".

mprse replied:
Thanks! Something like this https://github.com/ARMmbed/mbed-os/pull/7423/commits/523478792ac44a656b81501ebba9a38cb35f85be . 

I would like to have both versions ready since tomorrow I will be on the road to Croatia ;-)

deepikabhavnani replied:
I would vote for return instead of Asserts, in case of timeouts user should have control of software if they would like to perform cleanup and do some other activity instead of asserting?
Asserts are good idea for debug build but for stable OS we need other ways to deal with errors. 
I would prefer my OS to not break if I was not able to perform a specific task in one thread.

kjbracey replied:
The problem is that `mutex::lock()` failure is too fundamental and unexpected to realistically recover from. If your mutexes can't lock you have a real system problem, and it's not clear how you could recover.

Returning status from it is somewhat analogous to `*p = 3;` returning an error if `p` is an invalid pointer, rather than faulting.

And the error return would have to cascade up - the simplest `get_state()` method would need an error return if expected to "recover" from internal mutex lock failure.

See also the JIRA and issues linked in the description - this PR is the result of long previous discussion. No in-tree user is taking any action on mutex.lock() failure now. Either they ignore it, or they assert or error themselves. They have no "plan b" for a failure return.

deepikabhavnani replied:
Return status, will be useful for release profile. Users who are vigilant and do check error status will have advantage in case of release profiles for issues which are not caught in debug mode.

Consider a case where application has to spawn multiple threads one for each activity and due to overload (denial of service kind of activity) if mutex lock timeouts, they will not prefer whole system to crash or proceed with timed out lock. Recovery for that application will be to kill child threads (like connection lost).

> No in-tree user is taking any action on mutex.lock() failure now.

Does our cloud client application check return state?

> They have no "plan b" for a failure return.

I would consider that to be bug in user code, adding Assert is good for debug/develop profile. But for release build and system recovery of stable operating system error status is required.

deepikabhavnani replied:
> No in-tree user is taking any action on mutex.lock() failure now.

I see your point here, checked Mbed OS code, we do not check error status. I am still of view returning error status should not be removed even if we assert. Corner cases and bugs without error proper handling are difficult to debug in Release builds.

kjbracey replied:
Mutex locks cannot timeout via this API. This is the no-timeout `lock` API, for mutex claims that are expected to work. The separate `try_lock` API returns a boolean indicating whether it timed out or not.

An application of the type you hypothesise - one that expects to kill threads still holding mutexes if they timeout - would not be using this API to do it.

Part of the motivation here is to clearly distinguish `lock`, which always works, from `try_lock`, which may not.

> Does our cloud client application check return state?

Cloud client core doesn't use the RTOS, except via the in-tree ns_event_loop mechanisms.  The PAL stuff for the extension bits uses CMSIS-RTOS APIs, rather than the mbed OS wrapper, afaict.

> I am still of view returning error status should not be removed even if we assert. Corner cases and bugs without error proper handling are difficult to debug in Release builds.

Maybe it should be an `error` rather than an assert, so it still traps in release builds (as pointer failures do).

deepikabhavnani replied:
> Maybe it should be an error rather than an assert, so it still traps in release builds (as pointer failures do).

👍

bulislaw replied:
I would MBED_ASSERT should call error internally not die.

kjbracey commented on rtos/Mutex.h:
```diff
@@ -100,8 +97,9 @@ class Mutex : private mbed::NonCopyable<Mutex> {
                @a osErrorISR this function cannot be called from the interrupt service routine.
 
       @note You cannot call this function from ISR context.
+      @note This function asserts status of the lock operation (will not return in case of failure).
```
Sorry, you've gotten confused. `lock(millisec)` is wanted to be deprecated.

The change was to put back the `osStatus` on `lock()`, but say that use of it was deprecated. So this green bit goes onto `lock()`, and put back the deprecation on `lock(millisec)`.

AnotherButler commented on rtos/Mutex.h:
```diff
@@ -88,8 +97,10 @@ class Mutex : private mbed::NonCopyable<Mutex> {
 
       @note You cannot call this function from ISR context.
+      @note This function asserts status of the lock operation (will not return in case of failure).
+            Use of the return value is deprecated, and it's planned to become void in MBED 6.0.
```
Has this phrasing been approved by product? We should probably have deprecation warnings approved.
Also, please change "MBED" to "Mbed" unless there's a rebranding coming.

bulislaw replied:
Please remove reference to Mbed 6 and keep to other "standard" deprecation notes.

cmonr commented:
Device that timed out has been restarted.
/morph test

0xc0170 commented:
The CI results - should be rerun full build (rebased was done). 

I can still see here return type change (I would like to remove Breaking change label). We won't start any CI until this is approved.

cmonr commented:
@mprse Would you mind addressing @kjbracey-arm and @AnotherButler's comments before we restart CI?

@bulislaw Any comments?

cmonr commented:
@mprse This still needs comment work.

bulislaw reviewed (dismissed):
Please remove reference to mbed 6 and then it's approved.
Follow up work would be to change how assert works but I'll create a separate ticket.

kjbracey reviewed (changes requested):
The last commit messed the changes up, as I noted. `lock(millisec)` is supposed to be deprecated in favour of `try_lock(millisec)`.

Then the text that is currently attached to `lock(millisec)` about assertion and not using return value is supposed to be on `lock(void)`.

cmonr commented:
@0xc0170 @mprse @bulislaw @kjbracey-arm Is ie fair to say that this won't be making it into 5.10?

(Iirc, both @mprse and @bulislaw are away).

0xc0170 commented:
I believe it should get to 5.10. @mprse is back soon.

mprse committed:
Address review comments.

Perform the following changes:
- change definition of `void Mutex::lock(void)` to `osStatus Mutex::lock(void)`.
- change definition of `void Mutex::unlock()` to `osStatus Mutex::unlock()`.
- use MBED_ERROR1 macro to check the lock/unlock operation status.
- add notes in the description of lock/unlock functions: "This function asserts status of the lock/unlock operation (will not return in case of failure). Use of the return value is deprecated, as the return is expected to become void in the future.".
- modify/add description of the return value.
- remove reference to Mbed 6.
- make `lock(millisec)` deprecated in favour of lock(), trylock() and trylock_for() functions.

mprse commented:
Sorry for the delay. I'm back in office.

I rebased this PR, removed last commit and added fixes summarized by @kjbracey-arm here https://github.com/ARMmbed/mbed-os/pull/7423#pullrequestreview-148420523.

Also removed reference to mbed 6.0 https://github.com/ARMmbed/mbed-os/pull/7423#pullrequestreview-148410138.

kjbracey commented on rtos/Mutex.h:
```diff
@@ -78,8 +79,21 @@ class Mutex : private mbed::NonCopyable<Mutex> {
+
+      @note You cannot call this function from ISR context.
+      @note This function asserts status of the lock operation (will not return in case of failure).
+            Use of the return value is deprecated.
```
This version has no return value! The note was to be added because deepika and 0xc0170 wanted the `osStatus` return value retained on `lock()` for backwards source compatibility. I would add the clause ", as the return is expected to become `void` in future".

Also, it's not true that it won't return in case of failure in a release build, as deepika noted. I think we agree it should be changed to be an error/mbed_error call rather than assert.

With that change made, it will always be true that the function always returns osOK, or it doesn't return. And the doxygen for the return should be retained.

kjbracey commented on rtos/Mutex.h:
```diff
@@ -123,16 +138,12 @@ class Mutex : private mbed::NonCopyable<Mutex> {
       @note You cannot call this function from ISR context.
      */
-    osStatus unlock();
+    void unlock();
```
I think all the discussion about `lock` applies here too - return value is deprecated, always returns osOK, and doesn't return otherwise cos it calls error.

mprse commented:
Performed the following updates:
- change definition of `void Mutex::lock(void)` to `osStatus Mutex::lock(void)`.
- change definition of `void Mutex::unlock()` to `osStatus Mutex::unlock()`.
- use MBED_ERROR1 macro to check the lock/unlock operation status.
- add notes in the description of lock/unlock functions: "This function asserts status of the lock/unlock operation (will not return in case of failure). Use of the return value is deprecated, as the return is expected to become void in the future.".
- add description of the return value.
- remove reference to Mbed 6.
- make `lock(millisec)` deprecated in favour of `try_lock(millisec)`.

kjbracey commented on rtos/Mutex.h:
```diff
@@ -73,13 +75,28 @@ class Mutex : private mbed::NonCopyable<Mutex> {
+
+    /**
+      For backwards compatibility.
+      @deprecated Do not use this function. This function has been replaced with trylock_for and lock(void) functions.
```
The full set are `lock()`, `trylock()` and `trylock_for()`

mprse replied:
fixed.

kjbracey commented on rtos/Mutex.h:
```diff
@@ -89,7 +106,8 @@ class Mutex : private mbed::NonCopyable<Mutex> {
       @note You cannot call this function from ISR context.
      */
-    osStatus lock(uint32_t millisec=osWaitForever);
+    MBED_DEPRECATED_SINCE("mbed-os-5.10.0", "Replaced with try_lock(millisec) function")
```
`trylock()` and `trylock_for(millisec)`, and maybe mention `lock()` too. 

(That covers 0, finite and osWaitForever as the possible parameters).

mprse replied:
fixed.

kjbracey commented on rtos/Mutex.h:
```diff
@@ -73,13 +75,28 @@ class Mutex : private mbed::NonCopyable<Mutex> {
+               @a osOK the mutex has been obtained.
+
+      @note You cannot call this function from ISR context.
+      @note This function asserts status of the lock operation (will not return in case of failure).
```
Avoid the word "assert", as it's no longer an assert in the C sense. Maybe  "This function treats RTOS errors as fatal system errors, so can only return osOK".

mprse replied:
Fixed as suggested.

kjbracey commented on rtos/Mutex.h:
```diff
@@ -29,6 +29,8 @@
 #include "platform/NonCopyable.h"
 #include "platform/ScopedLock.h"
+#include "platform/mbed_toolchain.h"
+#include "platform/mbed_error.h"
```
These should probably be in the cpp file, not here?

mprse replied:
Removed since cpp already includes "mbed_error.h".

kjbracey commented on rtos/Mutex.h:
```diff
@@ -89,7 +106,8 @@ class Mutex : private mbed::NonCopyable<Mutex> {
      */
-    osStatus lock(uint32_t millisec=osWaitForever);
+    MBED_DEPRECATED_SINCE("mbed-os-5.10.0", "Replaced with try_lock(millisec) function")
+    osStatus lock(uint32_t millisec);
```
Now you've added the error to the implementation, osErrorParameter and osErrorISR are now impossible. I think that's fine, but remove them from doxygen and have a similar note here that RTOS errors other than the above are treated as fatal.

mprse replied:
Adapted the return values and added missing note.

kjbracey reviewed (commented):
Okay, that's basically correct now - minor adjustments, I think.

kjbracey reviewed (approved):
Looks good to me now.

0xc0170 commented:
/morph build

NirSonnenschein commented:
/morph export-build

NirSonnenschein commented:
/morph export-build

0xc0170 commented on rtos/Mutex.h:
```diff
@@ -73,23 +73,39 @@ class Mutex : private mbed::NonCopyable<Mutex> {
+            osErrorResource in case when millisec is 0 or osErrorTimeout if millisec is not osWaitForever.
      */
-    osStatus lock(uint32_t millisec=osWaitForever);
+    MBED_DEPRECATED_SINCE("mbed-os-5.10.0", "Replaced with lock(), trylock() and trylock_for() functions")
```
Look at exporters failing (this line there, as `mbed-os\rtos/Mutex.h(107): error:  #79: expected a type specifier`)

missing header file inclusion or what is this causing?

NirSonnenschein commented:
@mprse please see multiple failures in export.

mprse commented:
Looks like `#include "platform/mbed_toolchain.h"` is missing in `Mutex.h`. 
Added missing include.

It was already added before (https://github.com/ARMmbed/mbed-os/pull/7423/commits/ac2db7cc06f1f3b15f826a6b5821f8db477796b5)  and removed by mistake in the last commit while reverting wrong changes.

0xc0170 commented:
/morph build

adbridge commented:
Looks good.

MarceloSalazar commented:
FYI @daniel-cesarini-tridonic-com
